File: DSA/01-arrays-strings/arrays.js
/**
 * Arrays and Strings - JavaScript Implementation
 *
 * Topics covered:
 * - Array manipulation
 * - Two pointers
 * - Sliding window
 * - String operations
 */

// ============================================
// ARRAY BASICS
// ============================================

/**
 * Find maximum element in array
 * Time: O(n), Space: O(1)
 */
function findMax(arr) {
  if (arr.length === 0) return null;

  let max = arr[0];
  for (let i = 1; i < arr.length; i++) {
    if (arr[i] > max) {
      max = arr[i];
    }
  }
  return max;
}

/**
 * Reverse array in place
 * Time: O(n), Space: O(1)
 */
function reverseArray(arr) {
  let left = 0;
  let right = arr.length - 1;

  while (left < right) {
    [arr[left], arr[right]] = [arr[right], arr[left]];
    left++;
    right--;
  }

  return arr;
}

/**
 * Remove duplicates from sorted array
 * Time: O(n), Space: O(1)
 */
function removeDuplicates(nums) {
  if (nums.length === 0) return 0;

  let slow = 0;
  for (let fast = 1; fast < nums.length; fast++) {
    if (nums[fast] !== nums[slow]) {
      slow++;
      nums[slow] = nums[fast];
    }
  }

  return slow + 1;
}

/**
 * Rotate array to right by k positions
 * Time: O(n), Space: O(1)
 */
function rotateArray(nums, k) {
  k = k % nums.length;

  function reverse(start, end) {
    while (start < end) {
      [nums[start], nums[end]] = [nums[end], nums[start]];
      start++;
      end--;
    }
  }

  reverse(0, nums.length - 1);
  reverse(0, k - 1);
  reverse(k, nums.length - 1);

  return nums;
}

// ============================================
// TWO POINTERS
// ============================================

/**
 * Two Sum II - Input array is sorted
 * Time: O(n), Space: O(1)
 */
function twoSumSorted(numbers, target) {
  let left = 0;
  let right = numbers.length - 1;

  while (left < right) {
    const sum = numbers[left] + numbers[right];

    if (sum === target) {
      return [left, right];
    } else if (sum < target) {
      left++;
    } else {
      right--;
    }
  }

  return null;
}

/**
 * Container With Most Water
 * Time: O(n), Space: O(1)
 */
function maxArea(height) {
  let left = 0;
  let right = height.length - 1;
  let maxWater = 0;

  while (left < right) {
    const width = right - left;
    const currentHeight = Math.min(height[left], height[right]);
    const area = width * currentHeight;

    maxWater = Math.max(maxWater, area);

    if (height[left] < height[right]) {
      left++;
    } else {
      right--;
    }
  }

  return maxWater;
}

/**
 * Three Sum
 * Time: O(n²), Space: O(1)
 */
function threeSum(nums) {
  const result = [];
  nums.sort((a, b) => a - b);

  for (let i = 0; i < nums.length - 2; i++) {
    if (i > 0 && nums[i] === nums[i - 1]) continue;

    let left = i + 1;
    let right = nums.length - 1;
    const target = -nums[i];

    while (left < right) {
      const sum = nums[left] + nums[right];

      if (sum === target) {
        result.push([nums[i], nums[left], nums[right]]);

        while (left < right && nums[left] === nums[left + 1]) left++;
        while (left < right && nums[right] === nums[right - 1]) right--;

        left++;
        right--;
      } else if (sum < target) {
        left++;
      } else {
        right--;
      }
    }
  }

  return result;
}

/**
 * Valid Palindrome
 * Time: O(n), Space: O(1)
 */
function isPalindrome(s) {
  s = s.toLowerCase().replace(/[^a-z0-9]/g, "");

  let left = 0;
  let right = s.length - 1;

  while (left < right) {
    if (s[left] !== s[right]) {
      return false;
    }
    left++;
    right--;
  }

  return true;
}

// ============================================
// SLIDING WINDOW
// ============================================

/**
 * Maximum sum subarray of size k
 * Time: O(n), Space: O(1)
 */
function maxSumSubarray(arr, k) {
  if (arr.length < k) return null;

  let windowSum = 0;
  for (let i = 0; i < k; i++) {
    windowSum += arr[i];
  }

  let maxSum = windowSum;

  for (let i = k; i < arr.length; i++) {
    windowSum = windowSum - arr[i - k] + arr[i];
    maxSum = Math.max(maxSum, windowSum);
  }

  return maxSum;
}

/**
 * Longest Substring Without Repeating Characters
 * Time: O(n), Space: O(min(m, n))
 */
function lengthOfLongestSubstring(s) {
  const seen = new Map();
  let left = 0;
  let maxLength = 0;

  for (let right = 0; right < s.length; right++) {
    const char = s[right];

    if (seen.has(char) && seen.get(char) >= left) {
      left = seen.get(char) + 1;
    }

    seen.set(char, right);
    maxLength = Math.max(maxLength, right - left + 1);
  }

  return maxLength;
}

/**
 * Minimum Window Substring
 * Time: O(|S| + |T|), Space: O(|S| + |T|)
 */
function minWindow(s, t) {
  if (s.length === 0 || t.length === 0) return "";

  const need = new Map();
  for (let char of t) {
    need.set(char, (need.get(char) || 0) + 1);
  }

  let left = 0;
  let formed = 0;
  const required = need.size;
  const windowCounts = new Map();
  let ans = [Infinity, 0, 0];

  for (let right = 0; right < s.length; right++) {
    const char = s[right];
    windowCounts.set(char, (windowCounts.get(char) || 0) + 1);

    if (need.has(char) && windowCounts.get(char) === need.get(char)) {
      formed++;
    }

    while (left <= right && formed === required) {
      if (right - left + 1 < ans[0]) {
        ans = [right - left + 1, left, right];
      }

      const leftChar = s[left];
      windowCounts.set(leftChar, windowCounts.get(leftChar) - 1);

      if (
        need.has(leftChar) &&
        windowCounts.get(leftChar) < need.get(leftChar)
      ) {
        formed--;
      }

      left++;
    }
  }

  return ans[0] === Infinity ? "" : s.slice(ans[1], ans[2] + 1);
}

// ============================================
// STRING MANIPULATION
// ============================================

/**
 * Longest Palindromic Substring
 * Time: O(n²), Space: O(1)
 */
function longestPalindrome(s) {
  if (s.length < 2) return s;

  let start = 0;
  let maxLength = 1;

  function expandAroundCenter(left, right) {
    while (left >= 0 && right < s.length && s[left] === s[right]) {
      left--;
      right++;
    }
    return right - left - 1;
  }

  for (let i = 0; i < s.length; i++) {
    const len1 = expandAroundCenter(i, i);
    const len2 = expandAroundCenter(i, i + 1);
    const len = Math.max(len1, len2);

    if (len > maxLength) {
      maxLength = len;
      start = i - Math.floor((len - 1) / 2);
    }
  }

  return s.substring(start, start + maxLength);
}

/**
 * Group Anagrams
 * Time: O(n * k log k), Space: O(n * k)
 */
function groupAnagrams(strs) {
  const map = new Map();

  for (let str of strs) {
    const key = str.split("").sort().join("");

    if (!map.has(key)) {
      map.set(key, []);
    }
    map.get(key).push(str);
  }

  return Array.from(map.values());
}

/**
 * String to Integer (atoi)
 * Time: O(n), Space: O(1)
 */
function myAtoi(s) {
  const INT_MAX = 2 ** 31 - 1;
  const INT_MIN = -(2 ** 31);

  let i = 0;
  let sign = 1;
  let result = 0;

  while (i < s.length && s[i] === " ") i++;

  if (i < s.length && (s[i] === "+" || s[i] === "-")) {
    sign = s[i] === "+" ? 1 : -1;
    i++;
  }

  while (i < s.length && s[i] >= "0" && s[i] <= "9") {
    const digit = s[i].charCodeAt(0) - "0".charCodeAt(0);

    if (
      result > Math.floor(INT_MAX / 10) ||
      (result === Math.floor(INT_MAX / 10) && digit > INT_MAX % 10)
    ) {
      return sign === 1 ? INT_MAX : INT_MIN;
    }

    result = result * 10 + digit;
    i++;
  }

  return result * sign;
}

// ============================================
// MATRIX PROBLEMS
// ============================================

/**
 * Rotate Matrix 90 Degrees Clockwise
 * Time: O(n²), Space: O(1)
 */
function rotateMatrix(matrix) {
  const n = matrix.length;

  // Transpose
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      [matrix[i][j], matrix[j][i]] = [matrix[j][i], matrix[i][j]];
    }
  }

  // Reverse each row
  for (let i = 0; i < n; i++) {
    matrix[i].reverse();
  }

  return matrix;
}

/**
 * Spiral Matrix
 * Time: O(m*n), Space: O(1)
 */
function spiralOrder(matrix) {
  if (matrix.length === 0) return [];

  const result = [];
  let top = 0;
  let bottom = matrix.length - 1;
  let left = 0;
  let right = matrix[0].length - 1;

  while (top <= bottom && left <= right) {
    for (let i = left; i <= right; i++) {
      result.push(matrix[top][i]);
    }
    top++;

    for (let i = top; i <= bottom; i++) {
      result.push(matrix[i][right]);
    }
    right--;

    if (top <= bottom) {
      for (let i = right; i >= left; i--) {
        result.push(matrix[bottom][i]);
      }
      bottom--;
    }

    if (left <= right) {
      for (let i = bottom; i >= top; i--) {
        result.push(matrix[i][left]);
      }
      left++;
    }
  }

  return result;
}

/**
 * Set Matrix Zeroes
 * Time: O(m*n), Space: O(1)
 */
function setZeroes(matrix) {
  const m = matrix.length;
  const n = matrix[0].length;
  let firstRowZero = false;
  let firstColZero = false;

  // Check if first row should be zero
  for (let j = 0; j < n; j++) {
    if (matrix[0][j] === 0) {
      firstRowZero = true;
      break;
    }
  }

  // Check if first column should be zero
  for (let i = 0; i < m; i++) {
    if (matrix[i][0] === 0) {
      firstColZero = true;
      break;
    }
  }

  // Use first row and column as markers
  for (let i = 1; i < m; i++) {
    for (let j = 1; j < n; j++) {
      if (matrix[i][j] === 0) {
        matrix[i][0] = 0;
        matrix[0][j] = 0;
      }
    }
  }

  // Set zeros based on markers
  for (let i = 1; i < m; i++) {
    for (let j = 1; j < n; j++) {
      if (matrix[i][0] === 0 || matrix[0][j] === 0) {
        matrix[i][j] = 0;
      }
    }
  }

  // Handle first row
  if (firstRowZero) {
    for (let j = 0; j < n; j++) {
      matrix[0][j] = 0;
    }
  }

  // Handle first column
  if (firstColZero) {
    for (let i = 0; i < m; i++) {
      matrix[i][0] = 0;
    }
  }
}

// ============================================
// TESTS
// ============================================

function runTests() {
  console.log("=== Array Basics Tests ===");
  console.log("Find Max:", findMax([3, 1, 4, 1, 5, 9, 2])); // 9
  console.log("Reverse:", reverseArray([1, 2, 3, 4, 5])); // [5,4,3,2,1]
  console.log("Remove Duplicates:", removeDuplicates([1, 1, 2, 2, 3])); // 3
  console.log("Rotate Array:", rotateArray([1, 2, 3, 4, 5], 2)); // [4,5,1,2,3]

  console.log("\n=== Two Pointers Tests ===");
  console.log("Two Sum Sorted:", twoSumSorted([2, 7, 11, 15], 9)); // [0,1]
  console.log("Max Area:", maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7])); // 49
  console.log("Three Sum:", threeSum([-1, 0, 1, 2, -1, -4])); // [[-1,-1,2],[-1,0,1]]
  console.log("Is Palindrome:", isPalindrome("A man, a plan, a canal: Panama")); // true

  console.log("\n=== Sliding Window Tests ===");
  console.log("Max Sum Subarray:", maxSumSubarray([2, 1, 5, 1, 3, 2], 3)); // 9
  console.log("Longest Substring:", lengthOfLongestSubstring("abcabcbb")); // 3
  console.log("Min Window:", minWindow("ADOBECODEBANC", "ABC")); // "BANC"

  console.log("\n=== String Tests ===");
  console.log("Longest Palindrome:", longestPalindrome("babad")); // "bab" or "aba"
  console.log(
    "Group Anagrams:",
    groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"])
  );
  console.log("String to Int:", myAtoi("   -42")); // -42

  console.log("\n=== Matrix Tests ===");
  console.log(
    "Spiral Order:",
    spiralOrder([
      [1, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ])
  ); // [1,2,3,6,9,8,7,4,5]
}

// Run tests if this is the main module
if (require.main === module) {
  runTests();
}

module.exports = {
  findMax,
  reverseArray,
  removeDuplicates,
  rotateArray,
  twoSumSorted,
  maxArea,
  threeSum,
  isPalindrome,
  maxSumSubarray,
  lengthOfLongestSubstring,
  minWindow,
  longestPalindrome,
  groupAnagrams,
  myAtoi,
  rotateMatrix,
  spiralOrder,
  setZeroes,
};

File: DSA/01-arrays-strings/arrays.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

// ============================================
// ARRAY BASICS
// ============================================

// FindMax finds the maximum element in array
// Time: O(n), Space: O(1)
func FindMax(arr []int) (int, error) {
	if len(arr) == 0 {
		return 0, fmt.Errorf("empty array")
	}

	max := arr[0]
	for i := 1; i < len(arr); i++ {
		if arr[i] > max {
			max = arr[i]
		}
	}
	return max, nil
}

// ReverseArray reverses array in place
// Time: O(n), Space: O(1)
func ReverseArray(arr []int) []int {
	left, right := 0, len(arr)-1

	for left < right {
		arr[left], arr[right] = arr[right], arr[left]
		left++
		right--
	}

	return arr
}

// RemoveDuplicates removes duplicates from sorted array
// Time: O(n), Space: O(1)
func RemoveDuplicates(nums []int) int {
	if len(nums) == 0 {
		return 0
	}

	slow := 0
	for fast := 1; fast < len(nums); fast++ {
		if nums[fast] != nums[slow] {
			slow++
			nums[slow] = nums[fast]
		}
	}

	return slow + 1
}

// RotateArray rotates array to right by k positions
// Time: O(n), Space: O(1)
func RotateArray(nums []int, k int) []int {
	n := len(nums)
	k = k % n

	reverse := func(start, end int) {
		for start < end {
			nums[start], nums[end] = nums[end], nums[start]
			start++
			end--
		}
	}

	reverse(0, n-1)
	reverse(0, k-1)
	reverse(k, n-1)

	return nums
}

// ============================================
// TWO POINTERS
// ============================================

// TwoSumSorted finds two numbers that add up to target in sorted array
// Time: O(n), Space: O(1)
func TwoSumSorted(numbers []int, target int) []int {
	left, right := 0, len(numbers)-1

	for left < right {
		sum := numbers[left] + numbers[right]

		if sum == target {
			return []int{left, right}
		} else if sum < target {
			left++
		} else {
			right--
		}
	}

	return nil
}

// MaxArea finds the container with most water
// Time: O(n), Space: O(1)
func MaxArea(height []int) int {
	left, right := 0, len(height)-1
	maxWater := 0

	for left < right {
		width := right - left
		currentHeight := min(height[left], height[right])
		area := width * currentHeight

		maxWater = max(maxWater, area)

		if height[left] < height[right] {
			left++
		} else {
			right--
		}
	}

	return maxWater
}

// ThreeSum finds all triplets that sum to zero
// Time: O(n²), Space: O(1)
func ThreeSum(nums []int) [][]int {
	result := [][]int{}
	sort.Ints(nums)

	for i := 0; i < len(nums)-2; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		left := i + 1
		right := len(nums) - 1
		target := -nums[i]

		for left < right {
			sum := nums[left] + nums[right]

			if sum == target {
				result = append(result, []int{nums[i], nums[left], nums[right]})

				for left < right && nums[left] == nums[left+1] {
					left++
				}
				for left < right && nums[right] == nums[right-1] {
					right--
				}

				left++
				right--
			} else if sum < target {
				left++
			} else {
				right--
			}
		}
	}

	return result
}

// IsPalindrome checks if string is a valid palindrome
// Time: O(n), Space: O(1)
func IsPalindrome(s string) bool {
	s = strings.ToLower(s)

	left, right := 0, len(s)-1

	for left < right {
		// Skip non-alphanumeric characters
		for left < right && !isAlphanumeric(rune(s[left])) {
			left++
		}
		for left < right && !isAlphanumeric(rune(s[right])) {
			right--
		}

		if s[left] != s[right] {
			return false
		}

		left++
		right--
	}

	return true
}

func isAlphanumeric(c rune) bool {
	return (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9')
}

// ============================================
// SLIDING WINDOW
// ============================================

// MaxSumSubarray finds maximum sum of k consecutive elements
// Time: O(n), Space: O(1)
func MaxSumSubarray(arr []int, k int) (int, error) {
	if len(arr) < k {
		return 0, fmt.Errorf("array length less than k")
	}

	windowSum := 0
	for i := 0; i < k; i++ {
		windowSum += arr[i]
	}

	maxSum := windowSum

	for i := k; i < len(arr); i++ {
		windowSum = windowSum - arr[i-k] + arr[i]
		maxSum = max(maxSum, windowSum)
	}

	return maxSum, nil
}

// LengthOfLongestSubstring finds longest substring without repeating characters
// Time: O(n), Space: O(min(m, n))
func LengthOfLongestSubstring(s string) int {
	seen := make(map[byte]int)
	left := 0
	maxLength := 0

	for right := 0; right < len(s); right++ {
		char := s[right]

		if idx, ok := seen[char]; ok && idx >= left {
			left = idx + 1
		}

		seen[char] = right
		maxLength = max(maxLength, right-left+1)
	}

	return maxLength
}

// ============================================
// STRING MANIPULATION
// ============================================

// LongestPalindrome finds longest palindromic substring
// Time: O(n²), Space: O(1)
func LongestPalindrome(s string) string {
	if len(s) < 2 {
		return s
	}

	start := 0
	maxLength := 1

	expandAroundCenter := func(left, right int) int {
		for left >= 0 && right < len(s) && s[left] == s[right] {
			left--
			right++
		}
		return right - left - 1
	}

	for i := 0; i < len(s); i++ {
		len1 := expandAroundCenter(i, i)
		len2 := expandAroundCenter(i, i+1)
		length := max(len1, len2)

		if length > maxLength {
			maxLength = length
			start = i - (length-1)/2
		}
	}

	return s[start : start+maxLength]
}

// GroupAnagrams groups anagrams together
// Time: O(n * k log k), Space: O(n * k)
func GroupAnagrams(strs []string) [][]string {
	groups := make(map[string][]string)

	for _, str := range strs {
		// Sort string to use as key
		runes := []rune(str)
		sort.Slice(runes, func(i, j int) bool {
			return runes[i] < runes[j]
		})
		key := string(runes)

		groups[key] = append(groups[key], str)
	}

	result := make([][]string, 0, len(groups))
	for _, group := range groups {
		result = append(result, group)
	}

	return result
}

// ============================================
// MATRIX PROBLEMS
// ============================================

// RotateMatrix rotates matrix 90 degrees clockwise
// Time: O(n²), Space: O(1)
func RotateMatrix(matrix [][]int) [][]int {
	n := len(matrix)

	// Transpose
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
		}
	}

	// Reverse each row
	for i := 0; i < n; i++ {
		left, right := 0, n-1
		for left < right {
			matrix[i][left], matrix[i][right] = matrix[i][right], matrix[i][left]
			left++
			right--
		}
	}

	return matrix
}

// SpiralOrder returns matrix elements in spiral order
// Time: O(m*n), Space: O(1)
func SpiralOrder(matrix [][]int) []int {
	if len(matrix) == 0 {
		return []int{}
	}

	result := []int{}
	top, bottom := 0, len(matrix)-1
	left, right := 0, len(matrix[0])-1

	for top <= bottom && left <= right {
		// Traverse right
		for i := left; i <= right; i++ {
			result = append(result, matrix[top][i])
		}
		top++

		// Traverse down
		for i := top; i <= bottom; i++ {
			result = append(result, matrix[i][right])
		}
		right--

		// Traverse left
		if top <= bottom {
			for i := right; i >= left; i-- {
				result = append(result, matrix[bottom][i])
			}
			bottom--
		}

		// Traverse up
		if left <= right {
			for i := bottom; i >= top; i-- {
				result = append(result, matrix[i][left])
			}
			left++
		}
	}

	return result
}

// ============================================
// HELPER FUNCTIONS
// ============================================

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

// ============================================
// TESTS
// ============================================

func main() {
	fmt.Println("=== Array Basics Tests ===")
	max, _ := FindMax([]int{3, 1, 4, 1, 5, 9, 2})
	fmt.Printf("Find Max: %d\n", max) // 9

	fmt.Printf("Reverse: %v\n", ReverseArray([]int{1, 2, 3, 4, 5})) // [5,4,3,2,1]

	arr := []int{1, 1, 2, 2, 3}
	length := RemoveDuplicates(arr)
	fmt.Printf("Remove Duplicates: %d\n", length) // 3

	fmt.Printf("Rotate Array: %v\n", RotateArray([]int{1, 2, 3, 4, 5}, 2)) // [4,5,1,2,3]

	fmt.Println("\n=== Two Pointers Tests ===")
	fmt.Printf("Two Sum Sorted: %v\n", TwoSumSorted([]int{2, 7, 11, 15}, 9))          // [0,1]
	fmt.Printf("Max Area: %d\n", MaxArea([]int{1, 8, 6, 2, 5, 4, 8, 3, 7}))           // 49
	fmt.Printf("Three Sum: %v\n", ThreeSum([]int{-1, 0, 1, 2, -1, -4}))               // [[-1,-1,2],[-1,0,1]]
	fmt.Printf("Is Palindrome: %v\n", IsPalindrome("A man, a plan, a canal: Panama")) // true

	fmt.Println("\n=== Sliding Window Tests ===")
	maxSum, _ := MaxSumSubarray([]int{2, 1, 5, 1, 3, 2}, 3)
	fmt.Printf("Max Sum Subarray: %d\n", maxSum)                                // 9
	fmt.Printf("Longest Substring: %d\n", LengthOfLongestSubstring("abcabcbb")) // 3

	fmt.Println("\n=== String Tests ===")
	fmt.Printf("Longest Palindrome: %s\n", LongestPalindrome("babad")) // "bab" or "aba"
	fmt.Printf("Group Anagrams: %v\n", GroupAnagrams([]string{"eat", "tea", "tan", "ate", "nat", "bat"}))

	fmt.Println("\n=== Matrix Tests ===")
	fmt.Printf("Spiral Order: %v\n", SpiralOrder([][]int{{1, 2, 3}, {4, 5, 6}, {7, 8, 9}})) // [1,2,3,6,9,8,7,4,5]
}




